src/heaps/minmax_heap.jl

```diff
@@ -0,0 +1,261 @@
+
+
+################################################
+#
+# core implementation
+#
+################################################
+
+function _make_binary_minmax_heap(vt::Type{T}, xs) where {T}
+    n = length(xs)
+    valtree = copy(xs)
+    for i = n:-1:1
+        _minmax_heap_trickle_down!(valtree, i)
+    end
+    valtree
+end
+
+function _minmax_heap_bubble_up!(A::AbstractVector, i::Integer)
+    if level(i) % 2 == 0
+        # min level
+        if i > 1 && A[i] > A[hparent(i)]
+            # swap to parent and bubble up max
+            tmp = A[i]
+            A[i] = A[hparent(i)]
+            A[hparent(i)] = tmp
+            _minmax_heap_bubble_up!(A, hparent(i), Reverse)
+        else
+            # bubble up min
+            _minmax_heap_bubble_up!(A, i, Forward)
+        end
+        
+    else
+        # max level
+        if i > 1 && A[i] < A[hparent(i)]
+            # swap to parent and bubble up min
+            tmp = A[i]
+            A[i] = A[hparent(i)]
+            A[hparent(i)] = tmp
+            _minmax_heap_bubble_up!(A, hparent(i), Forward)
+        else
+            # bubble up max
+            _minmax_heap_bubble_up!(A, i, Reverse)
+        end
+    end
+   return 
+end
+
+function _minmax_heap_bubble_up!(A::AbstractVector, i::Integer, o::Ordering, x=A[i])
+    gparent = hparent(hparent(i))
+    if i > hparent(i) > gparent >= 1
+        # i has grandparent
+        if lt(o, x, A[gparent])
+            A[i] = A[gparent]
+            A[gparent] = x
+            _minmax_heap_bubble_up!(A, gparent, o)
+        end
+    end
+    return          
+end
+
+function _minmax_heap_trickle_down!(A::AbstractVector, i::Integer)
+    if level(i) % 2 == 0
+        _minmax_heap_trickle_down!(A, i, Forward)
+    else
+        _minmax_heap_trickle_down!(A, i, Reverse)
+    end
+    return
+end
+
+function _minmax_heap_trickle_down!(A::AbstractVector, i::Integer, o::Ordering, x=A[i])
+    N = length(A)
+    if lchild(i) ≤ N
+        # tuple (min(val, j1) max(val, j2))
+        ext = extrema((A[j], j) for j in descendants(N, i))
+        m = lt(o, ext[1], ext[2]) ? ext[1][2] : ext[2][2]
+
+        if m ≥ 4*i
+            # this is a grandchild
+            if lt(o, A[m], A[i])
+                A[i] = A[m]
+                A[m] = x
+                if lt(o, A[hparent(m)], A[m])
+                    t = A[m]
+                    A[m] = A[hparent(m)]
+                    A[hparent(m)] = t
+                end
+                _minmax_heap_trickle_down!(A, m, o)
+            end
+        else
+            if lt(o, A[m], A[i])
+                A[i] = A[m]
+                A[m] = x
+            end
+        end
+    end
+    return
+end
+                    
+################################################
+#
+# utilities
+#
+################################################
+
+@inline level(i) = round(Int, floor(log2(i)))
+@inline lchild(i) = 2*i
+@inline rchild(i) = 2*i+1
+@inline children(i) = [lchild(i), rchild(i)]
```

Comment:
return a `tuple` rather than a `Vector` if you don't plan on modifying the result.
Tuples are a generally more performant type for a number of reasons.